algorithms/color_refinement.py

```diff
@@ -0,0 +1,69 @@
+from supporting_components.graph import Graph, Vertex
+from typing import List
+
+
+def color_refinement(G: "Graph"):
+    """
+    Assuming there is no initial coloring
+    :param G: Graph to be colored
+    :return: Colored graph
+    """
+
+    colors, max_colornum = __initialize_colors(G)
+
+    has_changed = True
+    while has_changed:
+        for color_num, vertices in colors.copy().items():
+            if has_changed:
+                max_colornum += 1
+                has_changed = False
+
+            if len(vertices) == 1:
+                continue
+
+            has_changed = __colorgroup_refinement(colors, vertices, max_colornum)
```

Comment:
Volgens mij gaat dit nog niet helemaal goed nu. Als na de eerste kleuring de laatste groep van meer dan 1 vertex niets in veranderd, dan gaat hij uit de for loop met `has_changed` op `False` en dan gaat hij uit de while loop terwijl je eigenlijk nog niet klaar bent. Ik denk dat we hier twee `has_changed` variabelen moeten maken, eentje voor het ophogen van `max_colornum` en eentje die bijhoudt of er in deze ronde van de for loop iets veranderd is en dan moet die in de while loop staan.

Reply:
Dat komt denk ik doordat wij checken of die max_colornum veranderd moest worden, dat heb ik er voor nu uitgehaald omdat het eigenlijk ook niet nodig is. Wat we nu dan alleen krijgen is 1) Dat er gaten in de colornums komen die we gebruiken en 2) Als het aantal loops die hij moet maken erg groot is worden die colornums ook heel groot